annotations/service_test.go

```diff
@@ -0,0 +1,950 @@
+					require.NoError(t, err)
+					assertStoredAnnotations(t, tt.wantList, list)
+				})
+			}
+		})
+	})
+
+	t.Run("delete a single annotation by id", func(t *testing.T) {
+		svc, clean := newTestService(t)
+		defer clean(t)
+		ans := populateAnnotationsData(t, svc)
+
+		tests := []struct {
+			name         string
+			id           platform.ID
+			shouldDelete bool
+		}{
+			{
+				"has the correct error if not found",
+				idGen.ID(),
+				false,
+			},
+			{
+				"deletes the first one by id",
+				ans[0].ID,
+				true,
+			},
+			{
+				"deletes the second one by id",
+				ans[1].ID,
+				true,
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				err := svc.DeleteAnnotation(ctx, tt.id)
+
+				if tt.shouldDelete {
+					require.NoError(t, err)
+				} else {
+					require.Equal(t, errAnnotationNotFound, err)
+				}
+
+				got, err := svc.GetAnnotation(ctx, tt.id)
+				require.Equal(t, errAnnotationNotFound, err)
+				require.Nil(t, got)
+			})
+		}
+	})
+
+	t.Run("update a single annotation by id", func(t *testing.T) {
+		svc, clean := newTestService(t)
+		defer clean(t)
+		ans := populateAnnotationsData(t, svc)
+
+		updatedTime := time.Time{}.Add(time.Minute)
+
+		tests := []struct {
+			name    string
+			id      platform.ID
+			update  influxdb.AnnotationCreate
+			wantErr error
+		}{
+			{
+				"has the correct error if not found",
+				idGen.ID(),
+				influxdb.AnnotationCreate{
+					StreamTag: "updated tag",
+					Summary:   "updated summary",
+					Message:   "updated message",
+					Stickers:  map[string]string{"updated": "sticker"},
+					EndTime:   &updatedTime,
+					StartTime: &updatedTime,
+				},
+				errAnnotationNotFound,
+			},
+			{
+				"updates the first one by id",
+				ans[0].ID,
+				influxdb.AnnotationCreate{
+					StreamTag: "updated tag",
+					Summary:   "updated summary",
+					Message:   "updated message",
+					Stickers:  map[string]string{"updated": "sticker"},
+					EndTime:   &updatedTime,
+					StartTime: &updatedTime,
+				},
+				nil,
+			},
+			{
+				"updates the second one by id",
+				ans[1].ID,
+				influxdb.AnnotationCreate{
+					StreamTag: "updated tag2",
+					Summary:   "updated summary2",
+					Message:   "updated message2",
+					Stickers:  map[string]string{"updated2": "sticker2"},
+					EndTime:   &updatedTime,
+					StartTime: &updatedTime,
+				},
+				nil,
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				want := &influxdb.AnnotationEvent{ID: tt.id, AnnotationCreate: tt.update}
+				if tt.wantErr != nil {
+					want = nil
+				}
+
+				got, err := svc.UpdateAnnotation(ctx, tt.id, tt.update)
+				require.Equal(t, tt.wantErr, err)
+				require.Equal(t, want, got)
+
+				if tt.wantErr == nil {
+					new, err := svc.GetAnnotation(ctx, tt.id)
+					require.NoError(t, err)
+					e, err := new.ToEvent()
+					require.NoError(t, err)
+					require.Equal(t, got, e)
+				}
+			})
+		}
+	})
+
```

Comment:
I might have missed them somewhere in here, but can we test to be sure that renaming a stream that's referenced by some number of annotations works? Where "works" means:
1. The update request succeeds
2. Querying the annotations shows the updated stream name

I think that'd hit both of my concerns in the DB schema about the PK and tracking the stream name in multiple places.

Reply:
Good idea, that kind of operation definitely wasn't possible with the existing schema.

I got rid of the extra stream name column in the annotations table and revamped everything to work with joins. You can see the test I set up as you described in the latest commit.

I also found that some updates were needed to the `UpdateAnnotation` as part of this. You will now be able to update an annotation's stream reference by providing to a stream, which is either re-linked via the foreign key or created if needed to set the foreign key. Previously it would just change the stream tag value and not update the foreign key. De-normalization FTW!